services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,531 @@
+                            DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())));
+                }
+            });
+        }
+    }
+
+    /**
+     * Fetches tenant by subject DN.
+     *
+     * @param subjectDn the subject DN.
+     * @return A Future Map consisting of pair of a tenant id and a versioned tenant.
+     */
+    private Future<Map.Entry<String, Versioned<Tenant>>> getByCa(final X500Principal subjectDn) {
+
+        if (subjectDn == null) {
+            return Future.failedFuture("missing subject DN");
+        } else {
+            final JsonObject findTenantQuery = new MongoDbDocumentBuilder()
+                    .withCa("tenant", subjectDn.getName())
+                    .document();
+
+            final Promise<JsonObject> tenantsFound = Promise.promise();
+            mongoClient.findOne(getConfig().getCollectionName(), findTenantQuery, new JsonObject(), tenantsFound);
+            return tenantsFound.future().compose(successTenantsFound -> {
+                if (successTenantsFound == null || successTenantsFound.size() == 0) {
+                    return Future.succeededFuture(null);
+                }
+                final TenantDto tenantDtoFound = successTenantsFound.mapTo(TenantDto.class);
+                return Future.succeededFuture(
+                        new AbstractMap.SimpleEntry<>(tenantDtoFound.getTenantId(),
+                                new Versioned<>(tenantDtoFound.getVersion(),
+                                        tenantDtoFound.getTenant())));
+            });
+        }
+    }
+
+    @Override
+    public Future<OperationResult<Tenant>> readTenant(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+
+        return processReadTenant(tenantId, span)
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Tenant>> processReadTenant(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+
+        return findTenant(tenantId)
+                .compose(tenantDto -> Future.succeededFuture(OperationResult.ok(
+                        HttpURLConnection.HTTP_OK,
+                        tenantDto.getTenant(),
+                        Optional.ofNullable(DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())),
+                        Optional.ofNullable(tenantDto.getVersion()))));
+
+    }
+
+    /**
+     * Gets a tenant dto by tenant id.
+     *
+     * @param tenantId the tenant id.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future with a {@link TenantDto} be provided if found.
+     *         Otherwise the future will fail with
+     *         a {@link ClientErrorException} with {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     * @throws NullPointerException if the parameters is {@code null}.
```

Comment:
There is only one parameter,, how about changing to `parameter` instead of `parameters`.